File: go/problem_5.go
package main

import "fmt"

// cons constructs a pair encoded as a closure. car and cdr extract the first
// and second elements by passing selector functions into the pair.
func cons(a, b int) func(func(int, int) int) int {
	return func(f func(int, int) int) int {
		return f(a, b)
	}
}

// car returns the first element of a pair.
func car(pair func(func(int, int) int) int) int {
	return pair(func(a, _ int) int { return a })
}

// cdr returns the second element of a pair.
func cdr(pair func(func(int, int) int) int) int {
	return pair(func(_, b int) int { return b })
}

func problem5() {
	fmt.Println("Problem 5:")
	if car(cons(3, 4)) != 3 { panic("assertion failed") }
	if cdr(cons(3, 4)) != 4 { panic("assertion failed") }
	if car(cons(0, -1)) != 0 { panic("assertion failed") }
	if cdr(cons(0, -1)) != -1 { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}

File: lua/problem_5.lua
-- cons constructs a pair encoded as a closure. car and cdr extract the first
-- and second elements by passing selector functions into the pair.
local function cons(a, b)
    return function(f) return f(a, b) end
end

-- car returns the first element of a pair.
local function car(pair)
    return pair(function(a, _) return a end)
end

-- cdr returns the second element of a pair.
local function cdr(pair)
    return pair(function(_, b) return b end)
end

local function problem5()
    print("Problem 5:")
    assert(car(cons(3, 4))  == 3,  "assertion failed")
    assert(cdr(cons(3, 4))  == 4,  "assertion failed")
    assert(car(cons(0, -1)) == 0,  "assertion failed")
    assert(cdr(cons(0, -1)) == -1, "assertion failed")
    print("All Tests Passed")
end

return { problem5 = problem5 }
